csrc/multidevice/cuda_p2p.cpp

```diff
@@ -0,0 +1,70 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <cuda_utils.h>
+#include <multidevice/cuda_p2p.h>
+
+namespace nvfuser {
+
+namespace get_zcopy {
+
+void recvPost(const P2pIpcHandle& ipc_handles, int64_t count, CUstream stream) {
+  // wait for sender to be ready
+  NVFUSER_CUDA_SAFE_CALL(cuStreamWaitValue32(
+      stream,
+      reinterpret_cast<CUdeviceptr>(ipc_handles.local().semaphore()),
+      (cuuint32_t)(IpcSemaphore::kInUse),
+      CU_STREAM_WAIT_VALUE_EQ));
+  // RDMA get the data from the sender
+  NVFUSER_CUDA_RT_SAFE_CALL(cudaMemcpyAsync(
+      ipc_handles.local().ptr(),
+      ipc_handles.peer().ptr(),
+      count,
+      cudaMemcpyDeviceToDevice,
+      stream));
+  // Signals completion to self
+  NVFUSER_CUDA_SAFE_CALL(cuStreamWriteValue32(
+      stream,
+      reinterpret_cast<CUdeviceptr>(ipc_handles.local().semaphore()),
+      (cuuint32_t)(IpcSemaphore::kReady),
+      CU_STREAM_WRITE_VALUE_DEFAULT));
+  // Signals completion to sender
+  NVFUSER_CUDA_SAFE_CALL(cuStreamWriteValue32(
+      stream,
+      reinterpret_cast<CUdeviceptr>(ipc_handles.peer().semaphore()),
+      (cuuint32_t)(IpcSemaphore::kReady),
+      CU_STREAM_WRITE_VALUE_DEFAULT));
+}
+
+void sendPost(const P2pIpcHandle& ipc_handles, CUstream stream) {
```

Comment:
I couldn't map this implementation to http://nv/eA. so can you clarify? I was expecting the first step of sendPost to write kReady?

Reply:
Your link send me to https://dlrequest/GroupID/Home/Index

Reply:
> I was expecting the first step of sendPost to write kReady?

the first step is indeed to write to the semaphore. It writes `kInUse`, to signal it is ready-to-receive, while `kReady` signals the default semaphore state before the p2p starts. Are you asking why the name "kInUse" and not another naming?

Reply:
Fixed the link -- the short link generated by nv/ had a `.` in it, confusing GitHub's markdown renderer.

Reply:
> the first step is indeed to write to the semaphore. It writes `kInUse`, to signal it is ready-to-receive, while `kReady` signals the default semaphore state before the p2p starts. Are you asking why the name "kInUse" and not another naming?

Yes. That answered my question. I wasn't sure about the difference between "ready-to-receive" in the text and `kReady` in the code. Will read the code again based on the new understanding...